CL2-6566-create-new-eventpage-container

luucvanderzee opened the pull request:
Created the new event page. Still have to do the event cards- these are currently shown as placeholders. There is also some logic to handle pagination (see `eventViewerUtils.ts`) that will be removed in the future, but that I just put in there to test if the pagination and 'No new events' illustration show up correctly. I will add unit tests once the card is implemented.

A few things I ran into:

1. Should we put the 'filter' icon in [cl2-component-library](https://github.com/CitizenLabDotCo/cl2-component-library/blob/master/src/components/Icon/index.tsx) to make it more re-usable?
2. I now copy-pasted the [Title](https://github.com/CitizenLabDotCo/citizenlab/blob/CL2-6566-create-new-eventpage-container/front/app/containers/EventPage/EventViewer/TopBar.tsx#L20) in the `TopBar` component partly from the [PagesShowPage](https://github.com/CitizenLabDotCo/citizenlab/blob/CL2-6566-create-new-eventpage-container/front/app/containers/PagesShowPage/index.tsx#L80) one. I noticed there are a lot of similar 'title' styled components floating around. Would it make sense to make a reusable component out of these? And should I do that in this PR or put it on Jira?

luucvanderzee committed:
created event page and route

luucvanderzee committed:
updated cl-component-library

luucvanderzee committed:
implemented skeleton of EventPage

luucvanderzee committed:
made ProjectFilterDropdown re-usable, and added it to EventViewer

luucvanderzee committed:
extracted SectionContainer for re-usability and fixed margins in EventPage

luucvanderzee committed:
[CL2-6566] adding fallback picture for when there are no events (WIP)

luucvanderzee committed:
[CL2-6566] added 'no upcoming events' illustration

luucvanderzee committed:
[CL2-6566] extracted pagination component and added pagination to EventViewer

luucvanderzee committed:
[CL2-6566] added filter icon

luucvanderzee committed:
[CL2-6566] wrote messages

luucvanderzee committed:
[CL2-6566] finetuned some of the spacing

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
[CL2-6566] fixed spacing of first placeholder

luucvanderzee committed:
[CL2-6566] Merge branch 'temp' into CL2-6566-create-new-eventpage-container

luucvanderzee committed:
[CL2-6566] removed stray comment

brentguf commented on front/app/containers/EventPage/index.tsx:
```diff
@@ -0,0 +1,59 @@
+  margin-right: auto;
+`;
+
+const EventViewerWithTopMargin = styled(EventViewer)`
```
For code maintainability, I would not add the styles in the component name. If you need to change the styles, you also need to change the name. I'd go with `StyledEventViewer`.

Edit: for an even better name, see below

brentguf commented on front/app/containers/EventPage/index.tsx:
```diff
@@ -0,0 +1,59 @@
+  return (
+    <>
+      <Helmet>
+        <title>Events</title>
```
'Events' still needs to be added via formatMessage

brentguf commented on front/app/containers/EventPage/index.tsx:
```diff
@@ -0,0 +1,59 @@
+    .fill(0)
+    .map((_, i) => i + 1);
+
+  const pastEvents = [];
```
I assume you know, but just a reminder that these still need to be changed :)

brentguf commented on front/app/containers/EventPage/index.tsx:
```diff
@@ -0,0 +1,59 @@
+            title={<FormattedMessage {...messages.pastEvents} />}
+            fallbackMessage={<FormattedMessage {...messages.noPastEvents} />}
+            events={pastEvents}
+          />
```
To make things clearer immediately, you could rename the components to `UpcomingEventsViewer` and `PastEventsViewer`.

brentguf commented on front/app/containers/EventPage/EventViewer/FilterIcon.svg:
```diff
@@ -0,0 +1,3 @@
+<svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg">
+<path d="M0.158703 1.2075C1.6737 3.15 4.4712 6.75 4.4712 6.75V11.25C4.4712 11.6625 4.8087 12 5.2212 12H6.7212C7.1337 12 7.4712 11.6625 7.4712 11.25V6.75C7.4712 6.75 10.2612 3.15 11.7762 1.2075C11.8622 1.09678 11.9153 0.964147 11.9296 0.824705C11.9438 0.685263 11.9187 0.544616 11.8569 0.418778C11.7952 0.292941 11.6993 0.18697 11.5803 0.112932C11.4613 0.0388949 11.3239 -0.000234027 11.1837 1.05303e ...
+</svg>
```
Should be added to cl2-component-library indeed. :)

brentguf commented on front/app/containers/EventPage/EventViewer/TopBar.tsx:
```diff
@@ -0,0 +1,60 @@
+
+      <ProjectFilterDropdownPositioner>
+        <FilterIcon src={filterIcon} />
+        <ProjectFilterDropdown title="Projects" onChange={console.log} />
```
Still change the onChange

brentguf commented on front/app/containers/EventPage/EventViewer/TopBar.tsx:
```diff
@@ -0,0 +1,60 @@
+      <Title>{title}</Title>
+
+      <ProjectFilterDropdownPositioner>
+        <FilterIcon src={filterIcon} />
```
The icon needs the ariaHidden prop because screen readers will pronounce this SVG (and we should make this the default so we don't need to add this all the time). Putting it high on my todo. Or, feel free to do it when you add the filter icon to the component library. Let me know what you do!

brentguf commented on front/app/containers/EventPage/EventViewer/TopBar.tsx:
```diff
@@ -0,0 +1,60 @@
+  margin-top: auto;
+`;
+
+const FilterIcon = styled.img`
```
You should use the Icon component from cl2-component-library for icons.

brentguf commented on front/app/containers/EventPage/EventViewer/TopBar.tsx:
```diff
@@ -0,0 +1,60 @@
+  padding-top: 0px;
+
+  ${media.smallerThanMaxTablet`
+    font-size: ${fontSizes.xxxl};
```
Still add px here: `font-size: ${fontSizes.xxxl}px;`

brentguf commented on front/app/containers/EventPage/EventViewer/index.tsx:
```diff
@@ -0,0 +1,101 @@
+
+const EVENTS_PER_PAGE = 10;
+
+const EventViewer = memo<Props>(
```
Would rename to EventsViewer (plural) as we're viewing (possibly) multiple events. If it would be 1 event (like the idea page, project page), I'd use the singular name.

brentguf commented on front/app/containers/EventPage/EventViewer/index.tsx:
```diff
@@ -0,0 +1,101 @@
+  height: 237px;
+  margin-top: ${({ first }) => (first ? '29px' : '39px')};
+  padding: 30px;
+  font-size: 25px;
```
Like you've done it other places, we should rather use `font-size: ${fontSizes.xxl}px;`

brentguf reviewed (changes requested):
To answer your second question: you could add this to the component library, but would indeed do it outside of this PR. :)

luucvanderzee committed:
[CL2-6566] requested changes: naming and missing px in css

Commit:
Translations updated by CI (extract-intl)

guilherme-andrade commented on front/app/containers/EventsPage/EventsViewer/TopBar.tsx:
```diff
@@ -0,0 +1,60 @@
+
+      <ProjectFilterDropdownPositioner>
+        <FilterIcon src={filterIcon} />
+        <ProjectFilterDropdown title="Projects" onChange={console.log} />
```
`console.log` intentional? Waiting for the api I guess right?

guilherme-andrade commented on front/app/containers/EventsPage/EventsViewer/eventsViewerUtils.ts:
```diff
@@ -0,0 +1,22 @@
+export function sliceEventsToPage(
+  events: number[],
+  currentPage: number,
+  eventsPerPage: number
+) {
+  const numberOfEvents = events.length;
+
+  if (numberOfEvents === 0) return [];
+  if (numberOfEvents <= 10) return events;
+
+  const startIndex = (currentPage - 1) * eventsPerPage;
+  const endIndex = startIndex + eventsPerPage;
+
+  return events.slice(startIndex, endIndex);
+}
+
+export function getNumberOfPages(
+  numberOfEvents: number,
+  eventsPerPage: number
+) {
+  return Math.ceil(numberOfEvents / eventsPerPage);
+}
```
Is this used to paginate? The api is the best place to handle this, otherwise we're not solving the issue of very heavy response sizes, and if we have the logic in 2 places, then we need to maintain it in both FE and BE.

guilherme-andrade commented on front/app/containers/EventsPage/EventsViewer/index.tsx:
```diff
@@ -0,0 +1,101 @@
+        {events.length > 10 && (
+          <StyledPagination
+            currentPage={currentPage}
+            totalPages={getNumberOfPages(events.length, EVENTS_PER_PAGE)}
```
Same comments on pagination, I think you'll get this data from the api response.

guilherme-andrade commented:
Hey @luucvanderzee, added a few comments relative to API interactions but you can ignore them if these pagination and upcoming/past filtering utils were just added to mock the api calls.

Other than that looks great!

luucvanderzee committed:
[CL2-6566] updated icon and used correct text colors

luucvanderzee committed:
[CL2-6566] rename event page -> events page everywhere

luucvanderzee commented:
Thanks for the reviews! I fixed all requested changes, and will change the code to mock the api calls after merging with the API branch.

A few more issues came up:

1. Extracting a `Title` component and using it in all places where it can be re-used is a bit too much for now. This will be taken up in the technical debt regarding the component library.
2. I changed the color of the title of the filter dropdown to the `theme.colorText`, but on hover it always becomes black. Is this ok?
3. Sometimes the filter-funnel icon acts weird and moves a bit closer to the text for no apparent reason. I don't know how to reproduce it, but usually it happens if you leave the page open for a while and go to another tab. Any idea why this might happen?

How it should be:
![image](https://user-images.githubusercontent.com/13822005/123789772-d05c1600-d8dd-11eb-932e-091d07cd8817.png)

What sometimes happens:
![image](https://user-images.githubusercontent.com/13822005/123789631-aa367600-d8dd-11eb-8446-8f964aa95fb6.png)

brentguf committed:
Merge branch 'epic/CL2-6557/events-page' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6566-create-new-eventpage-container

brentguf commented:
> 2. I changed the color of the title of the filter dropdown to the `theme.colorText`, but on hover it always becomes black. Is this ok?

You can add :hover styles and use darken, but I'm not sure what the conventions are here. My advice would be to look at similar cases and how it's done there.

> 3. Sometimes the filter-funnel icon acts weird and moves a bit closer to the text for no apparent reason. I don't know how to reproduce it, but usually it happens if you leave the page open for a while and go to another tab. Any idea why this might happen?

The good thing about coding that there's always a reason. 😬 So I'd keep looking into reproducing it. Could it be related to the transform you're performing on the icon? Have you tried with margin instead?

luucvanderzee committed:
[CL2-6566] filter dropdown now darkens on hover

luucvanderzee committed:
[CL2-6566] fixed filter dropdown/icon strange behavior

brentguf commented:
@luucvanderzee I removed my review request for now so you can tag me again if you've looked at the two issues above. I'll get notified when you assign me again to review.

luucvanderzee committed:
[CL2-6566] Merge branch 'CL2-6566-create-new-eventpage-container' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6566-create-new-eventpage-container

luucvanderzee commented:
@brentguf both issues should be fixed now!

brentguf commented on front/app/containers/EventsPage/EventsViewer/TopBar.tsx:
```diff
@@ -0,0 +1,69 @@
+      <ProjectFilterDropdownPositioner>
+        <FilterIcon name="filter-funnel" />
+        <ProjectFilterDropdown
+          title="Projects"
```
"Projects" still needs to be added via formatMessage here

brentguf commented on front/app/containers/EventsPage/EventsViewer/TopBar.tsx:
```diff
@@ -0,0 +1,69 @@
+        <FilterIcon name="filter-funnel" />
+        <ProjectFilterDropdown
+          title="Projects"
+          onChange={console.log}
```
onChange is still console.log. I guess you're aware but I'm just commeting in case you forgot.

luucvanderzee replied:
Yes, noted :+1:

brentguf commented on front/app/containers/EventsPage/EventsViewer/TopBar.tsx:
```diff
@@ -0,0 +1,69 @@
+      <Title>{title}</Title>
+
+      <ProjectFilterDropdownPositioner>
+        <FilterIcon name="filter-funnel" />
```
Icon should be aria-hidden. You should also test how this component works with a screen reader. For more info on that, check our docs on accessibility (or send me a message to help you out :-)).

luucvanderzee replied:
The `Icon` component sets `aria-hidden` to `true` by default. Also, like in the other PR, I don't know what screen reader to use on Ubuntu

brentguf commented on front/app/containers/EventsPage/UpcomingEventsViewer.tsx:
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
```
To nitpick some more, I would call this component just UpcomingEvents in fact. And the other one PastEvents. Would be in line with our other components, but not blocking.

brentguf commented on front/app/containers/EventsPage/UpcomingEventsViewer.tsx:
```diff
@@ -0,0 +1,21 @@
+import messages from './messages';
+
+interface Props {
+  upcomingEvents: number[];
```
I know this is mock data, but wouldn't pass events such a long way. I think they should be fetched inside UpcomingEvents and PastEvents via a parameter in the request.

luucvanderzee replied:
Completely agree, will change this in the next PR when I will hook up everything with the API

brentguf commented on front/app/containers/EventsPage/index.tsx:
```diff
@@ -0,0 +1,45 @@
+
+  return (
+    <>
+      <Helmet>
```
Would add the meta info in a separate component (see e.g. IdeaIndexMeta). You need some more info than only the title.

brentguf reviewed (changes requested):
Almost there. Still a few more changes needed and a few more suggestions. :-)

luucvanderzee committed:
[CL2-6566] fixed requested changes

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
[CL2-6566] fixed front lint

brentguf commented on front/app/components/PageMeta/index.tsx:
```diff
@@ -0,0 +1,73 @@
+  }
+);
+
+const IdeaMetaWithHoc = injectIntl<Props>(IdeaMeta);
```
Given you are creating a new component, I would use hooks instead of resource components.

brentguf reviewed (approved):
One more suggestion, nothing blocking 👍

luucvanderzee committed:
[CL2-6566] replaced resource components with hooks in PageMeta
